File: MemoryDemo/MemoryDemo.cpp
#include <iostream>

void print(int* data, int size) {
    for (int i = 0; i < size; ++i) {
        std::cout << "C++ value: " << data[i] << std::endl;
    }
}

int main() {
    int* data = new int[5]; // heap allocation
    for (int i = 0; i < 5; ++i) {
        data[i] = i + 1;
    }
    print(data, 5);
    delete[] data; // must manually deallocate
    return 0;
}

File: MemoryDemo/MemoryDemo.rs
fn main() {
    let data = Box::new(vec![1, 2, 3, 4, 5]); // heap allocation via Box
    print(&data); // borrowing
    // data is automatically freed here (ownership goes out of scope)
}

fn print(v: &Vec<i32>) {
    for i in v {
        println!("Rust value: {}", i);
    }
}